packages/grid/x-data-grid-pro/src/hooks/features/columnReorder/useGridColumnReorder.tsx

```diff
@@ -97,6 +98,69 @@ export const useGridColumnReorder = (
       });
 
       originColumnIndex.current = apiRef.current.getColumnIndex(params.field, false);
+
+      const draggingColumnGroupPath = apiRef.current.getColumnGroupPath(params.field);
+
+      const visibleColumnIndex = apiRef.current.getColumnIndex(params.field, true);
+      const visibleColumns = apiRef.current.getVisibleColumns();
+      const groupsLookup = apiRef.current.getAllGroupDetails();
+
+      // The limitingGroupId is the id of the group from which the dragged column should ne escape
+      let limitingGroupId: string | null = null;
+
+      draggingColumnGroupPath.forEach((groupId) => {
+        if (!groupsLookup[groupId]?.freeReordering) {
+          // Only consider group that are made of more than one column
+          if (
+            visibleColumnIndex > 0 &&
+            visibleColumns[visibleColumnIndex - 1].groupPath?.includes(groupId)
+          ) {
+            limitingGroupId = groupId;
+          } else if (
+            visibleColumnIndex + 1 < visibleColumns.length &&
+            visibleColumns[visibleColumnIndex + 1].groupPath?.includes(groupId)
+          ) {
+            limitingGroupId = groupId;
+          }
+        }
+      });
+
+      forbiddenIndexes.current = {};
```

Comment:
That's par may need more comments. The overall idea is to consider the array of columns without the one dragged, and see which destinations are forbidden.

This allows to compute fobidden indexed once (when drag start)